File: src/services/ranking/__tests__/attributes.test.ts
import {
  enduranceScore,
  hunterRank,
  hunterScore,
  limitingAttribute,
  physiqueScore,
  strengthScore,
} from '@/services/ranking/attributes';

describe('strengthScore', () => {
  it('averages ranked exercise scores, or null when none', () => {
    expect(strengthScore([])).toBeNull();
    expect(strengthScore([40, 60, 80])).toBe(60);
  });
});

describe('physiqueScore (provisional, non-punitive)', () => {
  it('is null without an assessment', () => {
    expect(physiqueScore(null)).toBeNull();
  });
  it('peaks a healthy body fat and blends with muscle development', () => {
    // bodyFat 12% (male) -> 100; SMM 36/80 = 45% -> ~91.7; mean -> ~95.8
    const score = physiqueScore({
      bodyFatPercent: 12,
      skeletalMuscleMassKg: 36,
      weightKg: 80,
      sex: 'male',
    });
    expect(score).toBeCloseTo(95.8, 1);
  });
  it('does not reward ever-lower body fat (very low scores below the healthy peak)', () => {
    const healthy = physiqueScore({
      bodyFatPercent: 12,
      skeletalMuscleMassKg: null,
      weightKg: null,
      sex: 'male',
    })!;
    const veryLow = physiqueScore({
      bodyFatPercent: 4,
      skeletalMuscleMassKg: null,
      weightKg: null,
      sex: 'male',
    })!;
    expect(veryLow).toBeLessThan(healthy);
  });
});

describe('enduranceScore', () => {
  it('maps weekly training minutes, null when no data', () => {
    expect(enduranceScore(null)).toBeNull();
    expect(enduranceScore(0)).toBe(0);
    expect(enduranceScore(150)).toBe(70);
  });
});

describe('hunterScore & rank (PLAN.txt §6.7)', () => {
  it('applies attribute weights', () => {
    // 80*.4 + 60*.3 + 40*.15 + 60*.15 = 65
    const score = hunterScore({ strength: 80, physique: 60, endurance: 40, discipline: 60 });
    expect(score).toBeCloseTo(65, 5);
    expect(hunterRank(score)).toBe('A');
  });
  it('renormalizes over available attributes (missing physique/endurance/discipline)', () => {
    expect(hunterScore({ strength: 80, physique: null, endurance: null, discipline: null })).toBe(
      80,
    );
  });
  it('is neutral when nothing is known', () => {
    expect(hunterScore({ strength: null, physique: null, endurance: null, discipline: null })).toBe(
      60,
    );
  });
});

describe('limitingAttribute', () => {
  it('returns the lowest available attribute', () => {
    expect(limitingAttribute({ strength: 80, physique: 60, endurance: 40, discipline: 60 })).toBe(
      'endurance',
    );
    expect(
      limitingAttribute({ strength: 80, physique: null, endurance: null, discipline: 70 }),
    ).toBe('discipline');
    expect(
      limitingAttribute({ strength: null, physique: null, endurance: null, discipline: null }),
    ).toBeNull();
  });
});

File: android-native/app/src/test/java/com/fitnessrpg/app/domain/ranking/AttributesTest.kt
package com.fitnessrpg.app.domain.ranking

import com.fitnessrpg.app.domain.rank.Rank
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class AttributesTest {

    @Test
    fun `strengthScore averages ranked exercise scores, or null when none`() {
        assertNull(strengthScore(emptyList()))
        assertEquals(60.0, strengthScore(listOf(40.0, 60.0, 80.0))!!, 1e-9)
    }

    @Test
    fun `physiqueScore is null without an assessment`() {
        assertNull(physiqueScore(null))
    }

    @Test
    fun `physiqueScore peaks a healthy body fat and blends with muscle development`() {
        // bodyFat 12% (male) -> 100; SMM 36/80 = 45% -> ~91.7; mean -> ~95.8
        val score = physiqueScore(PhysiqueInput(12.0, 36.0, 80.0, "male"))
        assertEquals(95.8, score!!, 0.05)
    }

    @Test
    fun `physiqueScore does not reward ever-lower body fat`() {
        val healthy = physiqueScore(PhysiqueInput(12.0, null, null, "male"))!!
        val veryLow = physiqueScore(PhysiqueInput(4.0, null, null, "male"))!!
        assertTrue(veryLow < healthy)
    }

    @Test
    fun `enduranceScore maps weekly training minutes, null when no data`() {
        assertNull(enduranceScore(null))
        assertEquals(0.0, enduranceScore(0.0)!!, 1e-9)
        assertEquals(70.0, enduranceScore(150.0)!!, 1e-9)
    }

    @Test
    fun `hunterScore applies attribute weights`() {
        // 80*.4 + 60*.3 + 40*.15 + 60*.15 = 65
        val score = hunterScore(HunterAttributes(80.0, 60.0, 40.0, 60.0))
        assertEquals(65.0, score, 1e-4)
        assertEquals(Rank.A, hunterRank(score))
    }

    @Test
    fun `hunterScore renormalizes over available attributes`() {
        assertEquals(80.0, hunterScore(HunterAttributes(80.0, null, null, null)), 1e-9)
    }

    @Test
    fun `hunterScore is neutral when nothing is known`() {
        assertEquals(60.0, hunterScore(HunterAttributes(null, null, null, null)), 1e-9)
    }

    @Test
    fun `limitingAttribute returns the lowest available attribute`() {
        assertEquals(
            AttributeName.ENDURANCE,
            limitingAttribute(HunterAttributes(80.0, 60.0, 40.0, 60.0)),
        )
        assertEquals(
            AttributeName.DISCIPLINE,
            limitingAttribute(HunterAttributes(80.0, null, null, 70.0)),
        )
        assertNull(limitingAttribute(HunterAttributes(null, null, null, null)))
    }
}
